design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,319 @@
+ 
+For a service such implementation will look like this:
+```php
+namespace Magento\ServiceBProxy;
+
+class BManagerProxy implements Magento\ServiceBApi\BManagerInterface
+{
+    ....
+    
+    public function save(BDataInterface $data): BDataInterface
+    {
+        $promise = $this->invoker->invoke(
+            \Magento\ServiceBApi\BManagerInterface::class,
+            'save',
+            [$data]
+        );
+        $result = null;
+        $promise->then(function ($serviceExecResult) use (&$result) { $result = $serviceExecResult; });
+        $promise->otherwise(function (InvokerException $exception) { throw $exception; });
+        $promise->wait();
+        
+        if ($result->getException()) {
+            throw $result->getException();
+        }
+        
+        return $result->getResult();
+    }
+    
+    ....
+}
+```
+
+Proxy services will use _InvokerInterface_ to call remote services. It will return a promise which will pass a service execution result
+to _then_ callback and an exception to _otherwise_ callback if something during the remote call went wrong (like a network error). The
+service execution result may contain the return value of a service or an exception it has thrown. The promise that invoker returns may
+be used in both synchronous and asynchronous ways to be able to work with both types of services. More details about the promise
+[here](../promises.md).
+ 
+Proxy DTOs will look like this:
+```php
+namespace Magento\ServiceBProxy;
+
+class BDataProxy implements BDataInterface
+{
+    private $id;
+    
+    private $name;
+    
+    public function __construct(string $name, ?string $id) {
+        $this->name = $name;
+        $this->id = $id;
+    }
+    
+    public function getId(): ?string
+    {
+        return $this->id;
+    }
+    
+    public function setId(string $id): void
+    {
+        $this->id = $id;
+    }
+    
+    public function getName(): string
+    {
+        return $this->name;
+    }
+    
+    public function setName(string $name): void
+    {
+        $this->name = $name;
+    }
+}
+```
+Usually DTOs implementations are models in real modules but we need a class to be initialized when client code communicates with a
+remote service and for that simple DTOs will suffice.
+ 
+#### The invoker
+The invoker will be a part of Magento\Framework and will be present in every node installation. Invoker's job is to find a remote node's
+address based on given class name and method, send the request, receive response and convert it to either requested service method return type
+or an exception if it was thrown by the remote service. It returns a promise to support asynchronous services and because
+the requests to remote services will be asynchronous (depends on implementation). More details on the invoker's API and SPI
+[here](invoker.md).
+ 
+#### Gateway
+For actually delivering remote service requests RESTful API will be used - it's an established way to execute requested
```

Comment:
REST probably is not the best protocol for communication between services. We also can't violate protocol specification.

I suggest to mention some pros and cons for each of the options (REST, gRPC, RPC + REST, GraphQL), say that we need to investigate more gRPC and decision will be make in separate proposal. You can copy and paste what I have below.

REST
Pros
* We already have framework that provides support for it

Cons
* Not suitable for communication between services because it has limitations on the set of primary function calls (read, replace, modify, delete)
* Has large size of the message

gRPC
Pros
* Specifically designed for communication between services
* Small size of the message
* Support of HTTP 2
* Client code generation support for many programming languages (need to confirm if it works for PHP)
* Potentially can save on network communication if services would be demons that invokes service application

Cons
* There are some support limitations in PHP (rapidly growing though)

RPC + REST
Pros
* No limitations on the set of primary function calls 

Cons
* Response code for all requests is 200, which makes it hard to cache
* No versioning
* No strict schema

GraphQL
Pros
* ?

Cons
* Has complication logic behind it

Reply:
Mentioning my  comments as discussed during the last meeting:

I do find REST with JSON easier to debug - you may be able to use the browser directly or tools like Postman and/or curl, wireshark etc...

However there is no doubt gRPC with protobuffs is more efficient in space and time.  If we do go this route few things to confirm:

- With protobuffs, what happens if field ordering is changed as new fields are added on one side first (client or server)?. JSON can easily deal with this, but not sure if binary protocols are that flexible.  
- As most of the payload data types might be string type (e.g. in catalog with fields like long description, sku etc..), will there still be size benefits with protobuffs?